### PR TITLE
Implement beforeAuthStateChanged() Auth middleware function

### PR DESCRIPTION
**Pull request by hsubox76**
Design doc: [go/firebase-auth-middleware](https://goto.google.com/firebase-auth-middleware)

### CODE REVIEW

File: packages/auth/src/model/public_types.ts [hunk 1]
```diff
@@ -254,6 +254,16 @@ export interface Auth {
     error?: ErrorFn,
     completed?: CompleteFn
   ): Unsubscribe;
+  /**
```
Comment (hsubox76) [inline packages/auth/src/model/public_types.ts:257]: @egilmorez Can we get a review of the comments here? They will go into the reference docs.

**Review by egilmorez — APPROVED** [review]
LG!  Small nit: we avoid future tense whenever present suffices.  Thanks!

File: packages/auth/src/model/public_types.ts [hunk 2]
```diff
@@ -254,6 +254,16 @@ export interface Auth {
...
+   * Adds a blocking callback that runs before an auth state change
+   * sets a new user.
+   *
+   * @param callback - callback triggered before new user value is set.
+   *   If this throws, it will block the user from being set.
```
[the remark this answers is not in the document]
Comment (hsubox76) [reply packages/auth/src/model/public_types.ts:262]: I will add this update to https://github.com/firebase/firebase-js-sdk/pull/6221 and merge that into here.

**Review by dwyfrequency — APPROVED** [review]
LGTM

**Pull request merged**